src/array/cpu/spmat_op_impl_coo.cc

```diff
@@ -0,0 +1,497 @@
+    IdType newid = 0;
+    for (int64_t i = 0; i < len; ++i) {
+      const IdType id = ids_data[i];
+      if (!Contains(id)) {
+        oldv2newv_[id] = newid++;
+        filter_[id & kFilterMask] = true;
+      }
+    }
+  }
+
+  // Return true if the given id is contained in this hashmap.
+  bool Contains(IdType id) const {
+    return filter_[id & kFilterMask] && oldv2newv_.count(id);
+  }
+
+  // Return the new id of the given id. If the given id is not contained
+  // in the hash map, returns the default_val instead.
+  IdType Map(IdType id, IdType default_val) const {
+    if (filter_[id & kFilterMask]) {
+      auto it = oldv2newv_.find(id);
+      return (it == oldv2newv_.end()) ? default_val : it->second;
+    } else {
+      return default_val;
+    }
+  }
+
+ private:
+  static constexpr int32_t kFilterMask = 0xFFFFFF;
+  static constexpr int32_t kFilterSize = kFilterMask + 1;
+  // This bitmap is used as a bloom filter to remove some lookups.
+  // Hashtable is very slow. Using bloom filter can significantly speed up lookups.
+  std::vector<bool> filter_;
+  // The hashmap from old vid to new vid
+  std::unordered_map<IdType, IdType> oldv2newv_;
+};
+
+struct PairHash {
+  template <class T1, class T2>
+  std::size_t operator() (const std::pair<T1, T2>& pair) const {
+    return std::hash<T1>()(pair.first) ^ std::hash<T2>()(pair.second);
+  }
+};
+
+inline bool CSRHasData(CSRMatrix csr) {
+  return csr.data.defined();
+}
+
+inline bool COOHasData(COOMatrix csr) {
+  return csr.data.defined();
+}
+}  // namespace
+
+///////////////////////////// COOIsNonZero /////////////////////////////
+
+template <DLDeviceType XPU, typename IdType>
+bool COOIsNonZero(COOMatrix coo, int64_t row, int64_t col) {
+  CHECK(row >= 0 && row < coo.num_rows) << "Invalid row index: " << row;
+  CHECK(col >= 0 && col < coo.num_cols) << "Invalid col index: " << col;
+  const IdType* coo_row_data = static_cast<IdType*>(coo.row->data);
+  const IdType* coo_col_data = static_cast<IdType*>(coo.col->data);
+  for (int64_t i = 0; i < coo.row->shape[0]; ++i) {
+    if (coo_row_data[i] == row && coo_col_data[i] == col)
+      return true;
+  }
+  return false;
+}
+
+template bool COOIsNonZero<kDLCPU, int32_t>(COOMatrix, int64_t, int64_t);
+template bool COOIsNonZero<kDLCPU, int64_t>(COOMatrix, int64_t, int64_t);
+
+template <DLDeviceType XPU, typename IdType>
+NDArray COOIsNonZero(COOMatrix coo, NDArray row, NDArray col) {
+  const auto rowlen = row->shape[0];
+  const auto collen = col->shape[0];
+  const auto rstlen = std::max(rowlen, collen);
+  NDArray rst = NDArray::Empty({rstlen}, row->dtype, row->ctx);
+  IdType* rst_data = static_cast<IdType*>(rst->data);
+  const IdType* row_data = static_cast<IdType*>(row->data);
+  const IdType* col_data = static_cast<IdType*>(col->data);
+  const int64_t row_stride = (rowlen == 1 && collen != 1) ? 0 : 1;
+  const int64_t col_stride = (collen == 1 && rowlen != 1) ? 0 : 1;
+  const int64_t kmax = std::max(rowlen, collen);
```

Comment:
Does rowlen always == collen here? row and col are pairs.

Reply:
Not necessarily; we allow broadcasting here (i.e. row and col can contain only one element).